src/hipoc/hipoc_program.cpp

```diff
@@ -59,34 +66,73 @@ namespace miopen {
 #if !MIOPEN_USE_COMGR
 namespace {
 
-inline std::string GetCodeObjectVersionOptionImpl()
+int DetectCodeObjectOptionSyntax()
 {
-    auto code_object_version = miopen::Value(MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_VERSION{});
+    auto syntax = miopen::Value(MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_OPTION{});
+    if(syntax > 4)
+    {
+        MIOPEN_LOG_E("Bad MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_OPTION, using default");
+        syntax = 0;
+    }
+
+    if(syntax == 0)
+    {
+        if(HipCompilerVersion() >= external_tool_version_t{4, 0, 999999}) /// \ todo
+            return 4;
+        else
+            return 1;
+    }
+    MIOPEN_LOG_I("MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_OPTION=" << syntax);
+    return syntax;
+}
+
+int DetectCodeObjectVersion()
+{
+    auto co_version = miopen::Value(MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_VERSION{});
     // Very basic syntax check:
-    if(code_object_version == 1 || code_object_version > 4)
+    if(co_version == 1 || co_version > 4)
     {
         MIOPEN_LOG_E("Bad MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_VERSION, using default");
-        code_object_version = 0;
+        co_version = 0;
     }
 
-    if(HipCompilerVersion() >= external_tool_version_t{4, 0, -1})
+    if(co_version == 0)
     {
-        if(code_object_version == 0) // Env.var is unset.
-            return {};               // Rely on compiler's default.
+        if(HipCompilerVersion() >= external_tool_version_t{4, 0, 999999}) /// \ todo
```

Comment:
This should be updated when the whole stack (HIP compiler + runtime) supports CO v4.